File: Specialization_Data_Structures_Algorithms_UC_San_Diego/Course2_Data_Structures/week1_Basic_Data_Structures/tree_height/tree-height.py
# python3

import sys, threading
sys.setrecursionlimit(10**7) # max depth of recursion
threading.stack_size(2**27)  # new thread will get stack of such size

class TreeHeight(object):
    def __init__(self):
        self._n = 0
        self._children = []#using children representation
        self._parent =[]
        self._root = 0
    def read(self):
        self._n = int(sys.stdin.readline())
        for _ in range(0, self._n):
            self._children.append([])
        self._parent = [int(num) for num in sys.stdin.readline().split()]
        #list(map(int, sys.stdin.readline().split()))
        for index,parent in enumerate(self._parent):
            if(parent == -1):
                self._root = index
            else:
                self._children[parent].append(index)
        
    def compute(self,root):
        maxHeight = 0
        for index in range(0,len(self._children[root])):
            maxHeight = max(maxHeight,self.compute(self._children[root][index]))
        return maxHeight + 1

    def compute_height(self):
        # Replace this code with a faster implementation
        """
        maxHeight = 0
        for vertex in range(self.n):
                height = 0
                i = vertex
                while i != -1:
                        height += 1
                        i = self.parent[i]
                maxHeight = max(maxHeight, height);
        return maxHeight;
        """
        return self.compute(self._root)
            

def main():
    tree = TreeHeight()
    tree.read()
    print(tree.compute_height())

threading.Thread(target=main).start()

File: Specialization_Data_Structures_Algorithms_UC_San_Diego/Course2_Data_Structures/week1_Basic_Data_Structures/tree_height/tree_height.java
import java.util.*;
import java.io.*;
public class tree_height {
    class FastScanner {
		StringTokenizer tok = new StringTokenizer("");
		BufferedReader in;

		FastScanner() {
			in = new BufferedReader(new InputStreamReader(System.in));
		}

		String next() throws IOException {
			while (!tok.hasMoreElements())
				tok = new StringTokenizer(in.readLine());
			return tok.nextToken();
		}
		int nextInt() throws IOException {
			return Integer.parseInt(next());
		}
	}

	public class TreeHeight {
		int n;
		//int parent[];
		ArrayList<ArrayList<Integer>> children;
		int root;
		
		void read() throws IOException {
			FastScanner in = new FastScanner();
			n = in.nextInt();
			children = new ArrayList<ArrayList<Integer>>(n);
			for(int i = 0;i< n;i++)
				children.add(new ArrayList<Integer>());
			//parent = new int[n];
			for (int i = 0; i < n; i++) {

				int parent = in.nextInt();
				if(parent == -1)
					root = i;
				else {

					//System.out.println(children.get(parent).size());
					children.get(parent).add(i);
					//System.out.println("go");
				}
			}
		}
		int compute(int root){
			int max = 0;
			for (int i = 0;i < children.get(root).size();i++){
				max = Math.max(max,compute(children.get(root).get(i)));
			}
			return max + 1;
		}
		int computeHeight() {
                        // Replace this code with a faster implementation
			/*int maxHeight = 0;
			for (int vertex = 0; vertex < n; vertex++) {
				int height = 0;
				for (int i = vertex; i != -1; i = parent[i])
					height++;
				maxHeight = Math.max(maxHeight, height);
			}
			return maxHeight;
			*/
			return compute(root);
		}
	}

	static public void main(String[] args) throws IOException {
            new Thread(null, new Runnable() {
                    public void run() {
                        try {
                            new tree_height().run();
                        } catch (IOException e) {
                        }
                    }
                }, "1", 1 << 26).start();
	}
	public void run() throws IOException {
		TreeHeight tree = new TreeHeight();
		tree.read();
		System.out.println(tree.computeHeight());
	}
}
